kong/router/utils.lua

```diff
@@ -234,6 +234,131 @@ local function get_service_info(service)
          service_path
 end
 
+local phonehome_statistics
+do
+  local reports = require("kong.reports")
+  local nkeys = require("table.nkeys")
+  local log = ngx.log
+  local ERR = ngx.ERR
+
+  local byte = string.byte
+  local TILDE = byte("~")
+  local function is_regex_magic(path)
+    return byte(path) == TILDE
+  end
+
+  local empty_table = {}
+  -- reuse tables to avoid cost of creating tables and garbage collection
+  local protocols = {
+    http = 0, -- { "http", "https" },
+    stream = 0, -- { "tcp", "tls", "udp" },
+    tls_passthrough = 0, -- { "tls_passthrough" },
+    grpc = 0, -- { "grpc", "grpcs" },
+  }
+  local path_handlings = {
+    v0 = 0,
+    v1 = 0,
+  }
+  local route_report = {
+    flavor = "unknown",
+    paths = 0,
+    headers = 0,
+    routes = 0,
+    regex_routes = 0,
+    protocols = protocols,
+    path_handlings = path_handlings,
+  }
+
+  local function traditional_statistics(routes)
+    local paths_count = 0
+    local headers_count = 0
+    local regex_paths_count = 0
+    local http = 0
+    local stream = 0
+    local tls_passthrough = 0
+    local grpc = 0
+    local v0 = 0
+    local v1 = 0
+
+    for _, r in ipairs(routes) do
+      r = r.route
+      local paths = r.paths or empty_table
+      local headers = r.headers or empty_table
+
+      paths_count = paths_count + #paths
+      headers_count = headers_count + nkeys(headers)
+      for _, path in ipairs(paths) do
+        if is_regex_magic(path) then
+          regex_paths_count = regex_paths_count + 1
+          break
+        end
+      end
+
+      for _, protocol in ipairs(r.protocols or empty_table) do -- luacheck: ignore 512
+        if protocol == "http" or protocol == "https" then
+          http = http + 1
+
+        elseif protocol == "tcp" or protocol == "tls" or protocol == "udp" then
+          stream = stream + 1
+
+        elseif protocol == "tls_passthrough" then
+          tls_passthrough = tls_passthrough + 1
+
+        elseif protocol == "grpc" or protocol == "grpcs" then
+          grpc = grpc + 1
+
+        else
+          log(ERR, "unknown protocol: " .. protocol)
+        end
+        break
+      end
+
+      local path_handling = r.path_handling or "v0"
+      if path_handling == "v0" then
+        v0 = v0 + 1
+
+      elseif path_handling == "v1" then
+        v1 = v1 + 1
+      end
+    end
+
+    route_report.paths = paths_count
+    route_report.headers = headers_count
+    route_report.regex_routes = regex_paths_count
+    protocols.http = http
+    protocols.stream = stream
+    protocols.tls_passthrough = tls_passthrough
+    protocols.grpc = grpc
+    path_handlings.v0 = v0
+    path_handlings.v1 = v1
+  end
+
+  function phonehome_statistics(routes)
+    if not kong.configuration.anonymous_reports or ngx.worker.id() ~= 0 then
```

Comment:
We should cache `ngx.worker.id` in this `do end` block.

Reply:
The best we can do is to cache `ngx.worker` otherwise we will have id = 0 for all workers.

Reply:
My meaning is that we could use `local` to cache/localize the function `ngx.worker.id`.